man/smoother.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "SMOOTHER" "1" "" "smoother (vcflib)" "smoother (VCF transformation)"
.hy
.SH NAME
.PP
\f[B]smoother\f[R]
.SH SYNOPSIS
.PP
\f[B]smoother\f[R] \[en]format pFst \[en]file GPA.output.txt
.SH DESCRIPTION
.PP
smoothes is a method for window smoothing many of the GPAT++ formats.
.SH OPTIONS
.IP
.nf
\f[C]


      **smoother** averages a set of scores over a sliding genomic window.            
      **smoother** slides over genomic positions not the SNP indices. In other words  
      the number of scores within a window will not be constant. The last         
      window for each seqid can be smaller than the defined window size.          
      **smoother** automatically analyses different seqids separately.                
Output : 4 columns :     
     1. seqid            
     2. window start     
     2. window end       
     3. averaged score   

required: f,file     -- argument: a file created by GPAT++                           
required: o,format   -- argument: format of input file, case sensitive               
                              available format options:                                    
                                wcFst, pFst, bFst, iHS, xpEHH, abba-baba, col3             
optional: w,window   -- argument: size of genomic window in base pairs (default 5000)
optional: s,step     -- argument: window step size in base pairs (default 1000)      
optional: t,truncate -- flag    : end last window at last position (zero based)      

Type: transformation

\f[R]
.fi
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]not 0\f[R]
Failure
.SH SEE ALSO
.PP
\f[B]vcflib\f[R](1)
.SH OTHER
.SH LICENSE
.PP
Copyright 2011-2026 (C) Erik Garrison and vcflib contributors.
MIT licensed.
Copyright 2020-2026 (C) Pjotr Prins.
.SH AUTHORS
Erik Garrison and vcflib contributors.


RELEASE_NOTES.md
For contributions see
[contributors](https://github.com/vcflib/vcflib/graphs/contributors)
and
[commits](https://github.com/vcflib/vcflib/commits/master).

## ChangeLog ongoing

* Added guix-container-manifest.scm that allows building Singularity and Docker containers(!)

## ChangeLog v1.0.15 (20260319)

* Fix iterator usage in vcfkeepinfo and vcfremoveaberrantgenotypes (Jukka Matilainen)
* Fixes for vcfwave (Glenn Hickey)
* Added include files to build with vg
* Added pkgconfig support
* Fix wfa2lib builds and sent upstream
* Update zig build to 0.15, or later, and fix static build with zig for vcfcreatemulti
* Create tar ball with static binaries
* Fix building with git submodules only
* Download singularity and docker containers from https://files.genenetwork.org/software/vcflib/v1.0.15/
  md5sum:
  5768540e7bef2db1363e422bc37614c4  vcflib-1.0.15-docker.tar.gz # Docker
  680cd7675325561636e03274e679d195  vcflib-1.0.15-singularity.pack.gz # Singularity
  7f60bb29dfb498c1059a33e10e99ac02  vcflib-static-git-v1.0.15-2-g0d40ff1.tar.gz # Static binaries

## ChangeLog v1.0.14 (20250518)

* CMake: Fixed building static libs and shared libs (the previous version just compiled a static lib)
* Updated zig compile for 0.14 (MichaelK)
* Lots of refactoring (MichaelK)
* CMake: Improved linker speeds by moving some libs
* CMake: Improved markdown/pydoc tests, so we don't show the path to the binary
* guix.scm has multiple outputs including static binaries and a test mode
* guix.scm has procedure for building with local htslib and tabixpp only
* Added @RedPenguin100 Michael Kovaliov as a maintainer (welcome!)

## ChangeLog v1.0.13 (20250417)

* Zig support was added to Guix and Conda - and we expect it to be more common soon
* Addind release protocol to README - so I don't forget bumping version etc.
* Fixed vcfwave issues (see https://github.com/vcflib/vcflib/issues/416) and conda build (thanks Michael Kovaliov, aka @RedPenguin100)
* Fixed github actions (Michael)
* Lots of minor refactoring and cleanup, as well as a memory leak (Michael)
* tabix.hpp fixed and set `WFA_INCLUDE_DIRS` and `-DBUILD_SHARED_LIBS=ON` (thanks Michael Cho, aka @cho-m)
* Cmake fixes (thanks Adam Novak)
* Updated simde submodule
* Fix CMakeLists.txt not to try to install wfa-lib when it is a package
* Reverted on Canocalize extraction in Variant class because of downstream use, see https://github.com/vcflib/vcflib/issues/417

## ChangeLog v1.0.12 (20241122)

* Improved parsing of INFO and FORMAT lines by @jeizenga in https://github.com/vcflib/vcflib/pull/374
* Support external tabixpp by @adamnovak in https://github.com/vcflib/vcflib/pull/375
* Fix vcfwave.cpp by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/407
* Use regex to find the '..' between postions and replace it with '-' by @debbyku in https://github.com/vcflib/vcflib/pull/405
* Fix tag-fail long option that was overriden by tag-pass in vcffilter.cpp by @Gullumluvl in https://github.com/vcflib/vcflib/pull/404
* `vcfwave`: fix condition to avoid nullifying valid SNPs and MNPs by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/408
* Upgraded Zig support to 0.13.0 by @pjotrp
* Merged multichoose code into vcflib since no one else uses it
* Added sources for canonicalize too
* Moved Fasta.h (fastahack) dependencies from Variant.h into sources
* Updated multichoose and simde modules
* Improved vcfwave support

### New Contributors
* @jeizenga made their first contribution in https://github.com/vcflib/vcflib/pull/374
* @debbyku made their first contribution in https://github.com/vcflib/vcflib/pull/405
* @Gullumluvl made their first contribution in https://github.com/vcflib/vcflib/pull/404

## ChangeLog v1.0.11-pre ()

+ Stopped vendoring wfa2lib by default - so now the Debian build command is
  `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DZIG=OFF ..`


## ChangeLog v1.0.10 (20240420)

Vcflib maintenance release

+ Fixed vcfwave bugs - thanks Andrea Guarracino!
+ Fixed c++17 compatability - thanks Alex Petty!
+ Fixed tabixpp script - thanks Blaise Li!
+ Improved for vcfwave the wfa2lib build, fixed running tests with upstream build update
+ Removed deprecated binary_function
+ Removed unused files from repo
+ Updated README

## ChangeLog v1.0.9 (20230211)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.8 (20230210)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.7 (20230207)

Vcflib maintenance release - mostly for including in Debian

+ Fixed regression discovered by garguantua_kerr and atille of Debian (thanks!)
+ Added note on bio-vcf in vcffilter doc
+ notes on vcfcreatemulti and backtracking
+ CMake: honour include(GNUInstallDirs) paths (I forgot)

## ChangeLog v1.0.6 (20230129)

Vcflib maintenance release - mostly for including in Debian

+ Fixed zig complaining about leaking memory
+ Added CMake Debian support with -DWFA_GITMODULE=OFF
+ Introduced CMake include(GNUInstallDirs)
+ Successfully built wfa2 using embedded CMakeLists.txt
+ Cleaned up CMakeLists.txt removing comments etc.
+ Reintroduced vcfcreatemulti in legacy mode when ZIG=OFF (for Debian)

## ChangeLog v1.0.5 (20230116)

Vcflib's first *Humpty Dumpty* release: [vcfcreatemulti](./doc/vcfcreatemulti.md) is the natural companion to [vcfwave](./doc/vcfwave.md).

Often variant callers are not perfect.
**vcfwave** with its companion tool **vcfcreatemulti** can take an existing VCF file that contains multiple complex overlapping and even nested alleles and, unlike Humpty Dumpty, take them apart and put them together again.
Thereby, hopefully, creating sane VCF output that is useful for analysis and getting rid of false positives.

We created these tools by including the state-of-the-art [biWFA](https://github.com/smarco/WFA2-lib) wavefront aligner.

src/smoother.cpp
/*
    vcflib C++ library for parsing and manipulating VCF files

    Copyright © 2010-2020 Erik Garrison
    Copyright © 2020      Pjotr Prins

    This software is published under the MIT License. See the LICENSE file.
*/

#include <iostream>
#include <fstream>
#include <getopt.h>
#include <map>
#include <list>
#include <vector>
#include <string>
#include <cmath>

#include "split.h"
#include "gpatInfo.hpp"

using namespace std;

struct opts{
  bool     truncate;
  string   format;
  long int step;
  long int size;
  int      seqid;
  int      pos  ;
  int      value;
};

struct score{
  long int position;
  double score;
};



void printHelp(void){
  cerr << endl << endl;
  cerr << "INFO: help" << endl;
  cerr << "INFO: description:" << endl;
  cerr << "smoothes is a method for window smoothing many of the GPAT++ formats." << endl << endl ;
  cerr << "      smoother averages a set of scores over a sliding genomic window.            " << endl;
  cerr << "      smoother slides over genomic positions not the SNP indices. In other words  " << endl;
  cerr << "      the number of scores within a window will not be constant. The last         " << endl;
  cerr << "      window for each seqid can be smaller than the defined window size.          " << endl;
  cerr << "      smoother automatically analyses different seqids separately.                " << endl;

  cerr << "Output : 4 columns :     "    << endl;
  cerr << "     1. seqid            "    << endl;
  cerr << "     2. window start     "    << endl;
  cerr << "     2. window end       "    << endl;
  cerr << "     3. averaged score   "    << endl  << endl;

  cerr << "INFO: usage: smoother --format pFst --file GPA.output.txt" << endl;
  cerr << endl;
  cerr << "INFO: required: f,file     -- argument: a file created by GPAT++                           " << endl;
  cerr << "INFO: required: o,format   -- argument: format of input file, case sensitive               " << endl;
  cerr << "                              available format options:                                    " << endl;
  cerr << "                                wcFst, pFst, bFst, iHS, xpEHH, abba-baba, col3             " << endl;
  cerr << "INFO: optional: w,window   -- argument: size of genomic window in base pairs (default 5000)" << endl;
  cerr << "INFO: optional: s,step     -- argument: window step size in base pairs (default 1000)      " << endl;
  cerr << "INFO: optional: t,truncate -- flag    : end last window at last position (zero based)      " << endl;
  cerr << endl << "Type: transformation" << endl << endl;
  printVersion();
  cerr << endl << endl;
}

double ngreater(const std::list<score> & rangeData, double val){

  double n = 0;


  for(const auto& elm : rangeData){
    if(elm.score >= val){
      n += 1;
    }
  }
  return n;
}


double windowAvg(const std::list<score>& rangeData){

  double n = 0;
  double s = 0;

  for(const auto& elm : rangeData){
    s += elm.score;
    n += 1;
  }
  return (s/n);
}

//calculation of Patterson's D statistic
double dStatistic(const std::list<score> & rangeData){

  double abba = 0;
  double baba = 0;
  double dstat ;
  for(const auto& elm : rangeData){
    if(elm.score == 0){ // means we have BABA locus
      baba += 1;
    }
    else{ // count towards ABBA locus
      abba += 1;
    }
  }
  dstat = (abba - baba) / (abba + baba ); // d-statistic implementation
  return (dstat);
}

void processSeqid(ifstream & file, const string& seqid, streampos offset, opts & opt){

  string line ;

  long int windowSize = opt.size;
  long int start = 0;
  long int end   = windowSize;

  list<score> windowDat;

  file.clear();

  file.seekg(offset);

  vector<string> sline;

  while(getline(file, line)){

    sline = split(line, '\t');
    score current ;
    if(seqid != sline[opt.seqid]){
      break;
    }
    current.position = atol( sline[opt.pos].c_str() );
    current.score    = atof( sline[opt.value].c_str() );

    if(opt.format == "iHS"){
      current.score = fabs(current.score);
    }



    // add in if abba-baba to process second score.


    if(current.position > end){

      double reportValue ;

      if(opt.format == "abba-baba"){
	reportValue = dStatistic(windowDat);
      }
      else{
	reportValue = windowAvg(windowDat);
      }
      // nan
      if( reportValue == reportValue){
	cout << seqid << "\t" << start << "\t" << end << "\t" << windowDat.size() << "\t" << reportValue;
	if(opt.format == "iHS"){
	  std::cout << "\t" << ngreater(windowDat, 2.5) ;
	}
	std::cout << std::endl;
      }


    }
    while(end < current.position){
      start += opt.step;
      end   += opt.step;
      while(windowDat.front().position < start && !windowDat.empty()){
	windowDat.pop_front();
      }
    }
    windowDat.push_back(current);
  }
  // add function for D-stat if abba-baba
  double finalMean = windowAvg(windowDat);

  if(opt.truncate && (finalMean == finalMean) ){
    cout << seqid << "\t" << start << "\t" << windowDat.back().position - 1 << "\t" << windowDat.size()  << "\t" << finalMean;

    if(opt.format == "iHS"){
      std::cout << "\t" << ngreater(windowDat, 2.5) ;
    }
    std::cout << std::endl;

  }
  else if(finalMean == finalMean){
    cout << seqid << "\t" << start << "\t" << end << "\t" << windowDat.size() << "\t" << finalMean;

    if(opt.format == "iHS"){
      std::cout << "\t" << ngreater(windowDat, 2.5) ;
    }
    std::cout << std::endl;

  }
  cerr << "INFO: smoother finished : " << seqid << endl;
}

int main(int argc, char** argv) {

  map<string, int> acceptableFormats;
  acceptableFormats["pFst"]  = 1;
  acceptableFormats["col3"]  = 1;
  acceptableFormats["bFst"]  = 1;
  acceptableFormats["wcFst"] = 1;
  acceptableFormats["xpEHH"] = 1;
  acceptableFormats["iHS"]   = 1;
  acceptableFormats["cqf"]   = 1;
  acceptableFormats["deltaAf"]   = 1;
  acceptableFormats["abba-baba"]   = 1;

  opts opt;
  opt.size     = 5000;
  opt.step     = 1000;
  opt.format   = "NA";
  opt.truncate = false;

  string filename = "NA";

  static struct option longopts[] =
    {
      {"version"   , 0, 0, 'v'},
      {"help"      , 0, 0, 'h'},
      {"file"      , 1, 0, 'f'},
      {"window"    , 1, 0, 'w'},
      {"step"      , 1, 0, 's'},
      {"format"    , 1, 0, 'o'},
      {"truncate"  , 0, 0, 't'},
      {0,0,0,0}
    };

  int index;
  int iarg=0;

  while(iarg != -1){
    iarg = getopt_long(argc, argv, "f:w:s:o:vht", longopts, &index);
    switch(iarg){
    case 't':
      {
	opt.truncate = true;
	break;
      }
    case 'h':
      {
	printHelp();
	return 0;
      }
    case 'v':
      {
	printVersion();
	return 0;
      }
    case 'f':
      {
	filename = optarg;
	cerr << "INFO: file : " << filename << endl;
	break;
      }
    case 's':
      {
	opt.step = atol(optarg);
	cerr << "INFO: step size : " << optarg << endl;
	break;
      }
    case 'w':
      {
	opt.size = atol(optarg);
	cerr << "INFO: step size : " << optarg << endl;
	break;
      }
    case 'o':
      {
	opt.format = optarg;
	cerr << "INFO: specified input format : " << optarg << endl;
	break;
      }
    }
  }
  if(filename == "NA"){
    cerr << "FATAL: file was not specified!" << endl << endl;
    printHelp();
    return 1;
  }

  if(acceptableFormats.find(opt.format) == acceptableFormats.end()){
    cerr << "FATAL: unacceptable input file format, see --format "  << endl << endl;
    printHelp();
    return 1;
  }
  if(opt.format == "deltaAf"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 4;
  }
  else if(opt.format == "abba-baba"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 2;
  }
  else if(opt.format == "col3"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 2;
  }
  else if(opt.format == "pFst"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 2;
  }
  else if (opt.format == "bFst"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 8;
  }
  else if (opt.format == "wcFst"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 4;
  }
  else if(opt.format == "cqf"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 5;
  }
  else if(opt.format == "xpEHH"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 5;
  }
  else if(opt.format == "iHS"){
    opt.seqid = 0;
    opt.pos   = 1;
    opt.value = 6;
  }

  else{
    cerr << "FATAL: input format flag not specified correctly : " << opt.format << endl;
    cerr << "INFO:  please use smoother --help" << endl;
    return 1;
  }

  ifstream ifs(filename.c_str());

  string currentSeqid = "NA";

  string line;

  map<string, streampos > seqidIndex;

  if(ifs){
    while(getline(ifs, line)){
      vector<string> sline = split(line, '\t');
      if(sline[opt.seqid] != currentSeqid){

	long int bline = ifs.tellg() ;
	bline -=  ( line.size() +1 );

	//	std::cerr << "INFO: seqid: " << sline[opt.seqid] << " tellg: " << bline << std::endl;

	map<string, streampos>::iterator it;

	if(seqidIndex.find(sline[opt.seqid]) != seqidIndex.end() ){
	  cerr << "FATAL: file is unsorted!" << endl;
	  return 1;
	}
	seqidIndex[sline[opt.seqid]] = bline;
	currentSeqid = sline[opt.seqid];
      }
    }
  }
  else{
    cerr << "FATAL: no lines -- or -- couldn't open file" << endl;
  }

  for(const auto&[first, second] : seqidIndex) {
    cerr << "INFO: processing seqid : "<< first << endl;
    processSeqid(ifs, first, second, opt);
  }

  ifs.close();
  cerr << "INFO: smoother has successfully finished" << endl;

  return 0;

}
